### ISSUE TITLE
Generic type inference for doesn't work when extending nullable type

### ISSUE THREAD

**Issue by HosseinYousefi** [dart-lang/sdk#57001]
This works fine:

```dart
class Foo {}

class Bar<T extends Foo?> {
  final T foo;
  Bar(this.foo);
  
  static Bar<T> bar<T extends Foo>(T foo) {
    return Bar(foo); // <-- No need to write Bar<T>(foo)
  }
}
```

but if `T` extends `Foo?` instead, it doesn't:

```dart
class Foo {}

class Bar<T extends Foo?> {
  final T foo;
  Bar(this.foo);
  
  static Bar<T> bar<T extends Foo?>(T foo) {
    return Bar(foo);
    //         ^^^
    // The argument type 'T' can't be assigned to the parameter type 'Never'.
  }
}
```

Is this expected behavior?

**Comment by eernstg** [comment]
Related to https://github.com/dart-lang/sdk/issues/56998?

**Comment by johnniwinther** [comment]
I don't think it's related. @chloestefantsova can you take a look?

**Comment by lrhn** [comment]
Looks like a bug, or a very undeseriable property of inference.

The `return Bar(foo);` should infer `return Bar<T>(foo);` from the context type (return type of function is `Bar<T>`), and then `foo` is a valid argument.
If you write `return Bar<T>(foo);` it works. For *some* reason inference fails to do the obvious thing with a context type of `Bar<T>`, and infers `Bar<Never>(foo)` instead.

**Comment by johnniwinther** [comment]
@chloestefantsova Have you had a chance to look at this?

**Comment by chloestefantsova** [comment]
Yes, sorry for the delayed response here. It looks like it's an undesirable property of inference, like @lrhn mentioned. More precisely, it's an undesirable property of the lower bound algorithm that yields `Never` as the lower bound of `X extends A?` and `A?`.

The following is the rough outline of what happens:

1. The type parameter `Bar.T` is constrained by `Bar.bar.T` as a result of the constraint gathering (see https://github.com/dart-lang/language/blob/main/resources/type-system/inference.md#subtype-constraint-generation). So far so good: this is exactly what we intuitively want.
2. Since the solution of the constraints for `Bar.T` doesn't contain the unknown type (`_`), it's almost ready to be declared the preliminary inference result for that variable. As the final step, we need to check if adding the bound as the constraint won't yield a better result (see https://github.com/dart-lang/language/blob/main/resources/type-system/inference.md#constraint-solution-for-a-set-of-type-variables).
3. Adding the bound means merging two type constraints: `Bar.T <# Bar.bar.T` and `Bar.T <# Foo?`. Since both constraints are constraints from above, merging them means computing their lower bound.
4. The lower bound algorithm, as described in https://github.com/dart-lang/language/blob/main/resources/type-system/upper-lower-bounds.md#lower-bounds, will match the pair of types `Bar.bar.T` and `Foo?` and execute the step `DOWN(T1, T2?) = S where S is DOWN(T1, T2)`, removing the `?` from `Foo?`.
5. Finally, the lower bound algorithm yields `Never` for the pair of types `Bar.bar.T` and `Foo`.

This is very unfortunate, since the actual greatest lower bound for `Bar.bar.T` and `Foo?` is `Bar.bar.T`, but the algorithm produces `Never` as its best effort.

Note that in the first example that doesn't happen because in that case the lower bound algorithm compares two types of the form `X extends A`, `A?`, and `DOWN(X extends A, A?) = DOWN(X extends A, A) = X`.

@eernstg WDYT?

**Comment by eernstg** [comment]
I came to exactly the same result as @lrhn, and it would definitely yield the tightest possible lower bound if any of those two rules (namely

> - **DOWN**(`T1`, `T2`) = `T1` if `T1` <: `T2`
> - **DOWN**(`T1`, `T2`) = `T2` if `T2` <: `T1`

) is used at an arbitrarily early time. So in that sense it would be sound to move these subtype rules to the top.

However, if the reordering causes performance to deteriorate then we should put them after as many of the other rules as possible (that's probably the reason why they are currently located very far down &mdash; deciding `<:` is expensive).

So maybe it's better to move the following two rules down such that they occur right before the subtype rules shown above:

> - **DOWN**(`T1?`, `T2`) = `S` where `S` is **DOWN**(`T1`, `T2`)
> - **DOWN**(`T1`, `T2?`) = `S` where `S` is **DOWN**(`T1`, `T2`)

It is easy to see that no types of the form `T?` can match the rules between these two locations, which means that it is easy to see that this change will not change the behavior of **DOWN** in any other way than the one we need: The subtype rules will no have a chance to kick in before the `(T1, T2?)` and `(T1?, T2)` rules.

**Comment by lrhn** [comment]
We need to do " if `T1` \<: `T2` and `T2` \<!: `T1`" to avoid choosing arbitrarily for mutual subtypes.

We could also split the rules for `?`, like:

> * **DOWN**(`T1?`, `T2?`) = `S?` where `S` = **DOWN**(`T1`, `T2`)
> * **DOWN**(`T1?`, `T2`) = **DOWN**(`T1`, `T2`) if `T2` is non-nullable
> * **DOWN**(`T1`, `T2?`) = **DOWN**(`T1`, `T2`) if `T1` is non-nullable

which is the safe part that they currently assume, then we need more rules to cover `T1?` vs a potentially nullable `T2`.
Maybe check type variables and `FutureOr`s before checking `?` types again.
(Are type variables covered by the **DOWN** spec at all?)

**Comment by eernstg** [comment]
I think we have those properties already because all these typing helper functions are specified to choose the first applicable case. So when we have checked whether the `(T1?, T2?)` rule can be applied, we know for sure in the subsequent rules that the operands aren't of the form `T1?` and `T2?`.

I still think it's enough (and that it is the least disruptive approach) to move the `(T1?, T2)` and `(T1, T2?)` rules down immediately after the subtype rules.
